subsys/bluetooth/host/conn.c

```diff
@@ -1153,6 +1158,104 @@ struct bt_conn *conn_lookup_handle(struct bt_conn *conns, size_t size,
 	return NULL;
 }
 
+#if defined(CONFIG_BT_CONN)
+/* We don't want the application to get a PHY update callback upon connection
+ * establishment on 2M PHY. Therefore we must prevent issuing LE Set PHY
+ * in this scenario.
+ */
+static bool skip_auto_phy_update_on_conn_establishment(struct bt_conn *conn)
+{
+#if defined(CONFIG_BT_USER_PHY_UPDATE)
+	if (IS_ENABLED(CONFIG_BT_AUTO_PHY_UPDATE) &&
+	    IS_ENABLED(CONFIG_BT_EXT_ADV) &&
+	    BT_DEV_FEAT_LE_EXT_ADV(bt_dev.le.features)) {
+		if (conn->le.phy.tx_phy == BT_HCI_LE_PHY_2M &&
+		    conn->le.phy.rx_phy == BT_HCI_LE_PHY_2M) {
+			return true;
+		}
+	}
+#else
+	ARG_UNUSED(conn);
+#endif /* defined(CONFIG_BT_USER_PHY_UPDATE) */
+
+	return false;
+}
+
+static void auto_initiated_procedures(struct k_work *work)
+{
+	int err;
+	struct bt_conn *conn = CONTAINER_OF(work, struct bt_conn, auto_initiated_procedures);
+
+	if (conn->state != BT_CONN_CONNECTED) {
+		/* It is possible that connection was disconnected directly from
+		 * connected callback so we must check state before doing
+		 * connection parameters update.
+		 */
+		goto exit;
+	}
+
+	if (!atomic_test_bit(conn->flags, BT_CONN_AUTO_FEATURE_EXCH) &&
+	    ((conn->role == BT_HCI_ROLE_CENTRAL) ||
+	     BT_FEAT_LE_PER_INIT_FEAT_XCHG(bt_dev.le.features))) {
+		err = bt_hci_le_read_remote_features(conn);
+		if (err) {
+			LOG_ERR("Failed read remote features (%d)", err);
+		}
+		if (conn->state != BT_CONN_CONNECTED) {
+			goto exit;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_REMOTE_VERSION) &&
+	    !atomic_test_bit(conn->flags, BT_CONN_AUTO_VERSION_INFO)) {
+		err = bt_hci_read_remote_version(conn);
+		if (err) {
+			LOG_ERR("Failed read remote version (%d)", err);
+		}
+		if (conn->state != BT_CONN_CONNECTED) {
+			goto exit;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_AUTO_PHY_UPDATE) &&
+	    BT_FEAT_LE_PHY_2M(bt_dev.le.features) &&
+	    !skip_auto_phy_update_on_conn_establishment(conn)) {
+		err = bt_le_set_phy(conn, 0U, BT_HCI_LE_PHY_PREFER_2M,
+				    BT_HCI_LE_PHY_PREFER_2M,
+				    BT_HCI_LE_PHY_CODED_ANY);
+		if (err) {
+			LOG_ERR("Failed LE Set PHY (%d)", err);
+		}
+		if (conn->state != BT_CONN_CONNECTED) {
+			goto exit;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_AUTO_DATA_LEN_UPDATE) &&
+	    BT_FEAT_LE_DLE(bt_dev.le.features)) {
+		if (bt_drv_quirk_no_auto_dle()) {
+			uint16_t tx_octets, tx_time;
+
+			err = bt_hci_le_read_max_data_len(&tx_octets, &tx_time);
+			if (!err) {
+				err = bt_le_set_data_len(conn,
+						tx_octets, tx_time);
+				if (err) {
+					LOG_ERR("Failed to set data len (%d)", err);
+				}
+			}
+		} else {
+			/* No need to auto-initiate DLE procedure.
+			 * It is done by the controller.
+			 */
+		}
+	}
+
+exit:
+	bt_conn_unref(conn);
```

Comment:
This can potentially destroy the currently running k_work object. Is that ok?

Reply:
good point. I have no idea if it's going to implode or not.

Reply:
It looks in work.c like the work item is touched after the handler completes. Better safe than sorry. I suggest we make the work item global, we iterate over all connected conns in the handler, and have a flag on each conn that marks the work done.

Reply:
Looking at what the workqueue implementation does after calling the callback this may actually be a problem (it's still accessing parts of the work struct, like `work->flags` after it).
https://github.com/zephyrproject-rtos/zephyr/blob/40414f72e9fcec733b0ad4d046c3c042a579bb94/kernel/work.c#L688-L703

Reply:
I wonder if this is already an issue with `deferred_work`:
https://github.com/zephyrproject-rtos/zephyr/blob/40414f72e9fcec733b0ad4d046c3c042a579bb94/subsys/bluetooth/host/conn.c#L2069-L2072

Reply:
would be pretty easy to test btw: add an assert in `notify_recycled_conn_slot` that we're not running in the syswq.

Reply:
One solution could be to do "slow" freeing of connections, i.e. when the refcount drops to zero there's a separate (independent of any connection object) `k_work` that's responsible for really freeing the object.

Reply:
> One solution could be to do "slow" freeing of connections, i.e. when the refcount drops to zero there's a separate (independent of any connection object) `k_work` that's responsible for really freeing the object.

Wouldn't that hinder advertisers to restart connectable advertising even further? 
Today we suggest to use a k_work in the disconnected callback, but if the stack then also does a k_work before actually freeing the connection, then the application's k_work needs to be schedules after the stack's k_work.

Reply:
we probably need to address the rootcause, which is the work item belonging to the `struct bt_conn`. I really don't like the `deferred_work` anyways. So refactoring `deferred_work` using @alwa-nordic 's proposition seems appropriate.

Reply:
> we probably need to address the rootcause, which is the work item belonging to the `struct bt_conn`. I really don't like the `deferred_work` anyways. So refactoring `deferred_work` using @alwa-nordic 's proposition seems appropriate.

Not really opposed to that either. It does sound a bit like we are starting to implement a tiny garbage collector if we have a work item that occasionally goes through all connection objects to finalize the free'ing :D But you are correct that free'ing an object that contains the work items that triggers the free is an issue

Reply:
I suggest to name it ZNGC _Zis is Not a Garbage Collector_ 🚮🙊

Reply:
> ZNGC Zis is Not a Garbage Collector

Better make it "ZNGC: ZNGC is Not a Garbage Collector" for additional recursiveness